File: TotoRan/newRootViewController.m
//
//  newRootViewController.m
//  TotoRan
//
//  Created by 大山 孝 on 2015/09/06.
//  Copyright (c) 2015年 raksam.com. All rights reserved.
//

#import "newRootViewController.h"

#define WIDTH self.view.bounds.size.width
#define HEIGHT self.view.bounds.size.height

#define LOGO_RECT CGRectMake(WIDTH / 2 - 50, (HEIGHT / 8) - 10, 100, 100)
#define BTNRECT_SINGLE CGRectMake(WIDTH / 2 - ((WIDTH / 2) / 2), (HEIGHT / 8) * 3 - 10, WIDTH / 2, HEIGHT / 10)
#define BTNRECT_MULTI CGRectMake(WIDTH / 2 - ((WIDTH / 2) / 2), (HEIGHT / 8) * 5 - 10, WIDTH / 2, HEIGHT / 10)
#define ENDDATE_LABEL_RECT CGRectMake(WIDTH / 2 - ((WIDTH * 0.75) / 2), (HEIGHT / 8) * 7 - 40, WIDTH * 0.75, 50)
#define ROOP_AN_COUNT 10000
#define ROOP_AN_MESSAGE1 @"何かしらの異常が発生しました\nアプリを再起動してください\n※Jリーグシーズンオフ中はデータがないので起動できません"
#define ROOP_AN_MESSAGE2 @"何かしらの異常が発生しました\nアプリを再起動してください"
#define NOT_KAISAI_MESSAGE @"現在開催中のtotoはありません"
#define FOUR_WIDTH 320.0
#define SIX_WIDTH 375.0
#define SIXPLUS_WIDTH 414.0
#define FIVE_HEIGHT 960
#define BOARD_WITH 3.0
#define FOUR_FONT 12
#define FIVE_FONT 14
#define SIX_FONT 16
#define SIXPLUS_FONT 18
#define SLEEP_TIME 3.0
#define WAKU_COUNT 13

enum {TOTO, BODDS};
enum {SINGLE = 101, MULTI};
enum {ZERO, ONE, TWO, THREE};


@interface newRootViewController ()
@end

@implementation newRootViewController
{
    BOOL abnomalFlg; //何かしらの異常がある場合は次画面への遷移を抑制する
    NSString *abnomalMessage; //何かしらの異常がある場合に次画面遷移を試みた場合に表示するメッセージ
    AppDelegate *appDelegate; //通信とDB失敗データの取得と各種データの格納用
    int fontSizeDispatch; //機種別フォントサイズ
}

//初期化
- (instancetype)init
{
    self = [super init];
    if (self) {
        abnomalFlg = NO;
        abnomalMessage = nil;
        fontSizeDispatch = ZERO;
    }
    return self;
}


- (void)viewDidLoad {
    [super viewDidLoad];

    //背景色は黄色で黒で縁取り ※黒で縁取ると黒いiPhoneだとただ画面がちっちゃく見えるだけ。。。
    self.view.backgroundColor = [UIColor yellowColor];
    
    //機種を画面サイズ幅で制御 4シリーズと5シリーズは320.0、6は375.0、6+は414.0　※6sシリーズを考慮すっぺ
    float widthSize = self.view.bounds.size.width;
    float HeightSize = self.view.bounds.size.height;
    if (widthSize <= FOUR_WIDTH) {
        if (HeightSize <= FIVE_HEIGHT) {
            fontSizeDispatch = FOUR_FONT;
        } else {
            fontSizeDispatch = FIVE_FONT;
        }
    } else if (widthSize >= SIX_WIDTH && SIXPLUS_WIDTH > widthSize) {
        fontSizeDispatch = SIX_FONT;
    } else {
        fontSizeDispatch = SIXPLUS_FONT;
    }
    
    //ロゴ
    UIImageView *logo = [[UIImageView alloc] initWithFrame:LOGO_RECT];
    logo.image = [UIImage imageNamed:@"TotoRanLogo.png"];
    [self.view addSubview:logo];
    
    //ボタン
    UIButton *singleBtn = [self makeButton:@"シングル選択" frame:BTNRECT_SINGLE tag:SINGLE];
    UIButton *multiBtn = [self makeButton:@"マルチ選択" frame:BTNRECT_MULTI tag:MULTI];
    [self.view addSubview:singleBtn];
    [self.view addSubview:multiBtn];
}

//ボタン生成メソッド
- (UIButton *)makeButton:(NSString *)titile frame:(CGRect)frame tag:(int)tag
{
    UIButton *btn = [[UIButton alloc] initWithFrame:frame];
    btn.tag = tag;
    [btn setTitleColor:[UIColor blackColor] forState:UIControlStateNormal];
    [btn setTitleColor:[UIColor lightGrayColor] forState:UIControlStateHighlighted];
    [btn setTitle:titile forState:UIControlStateNormal];
    [btn addTarget:self action:@selector(btnPush:) forControlEvents:UIControlEventTouchUpInside];
    [btn.titleLabel setFont:[UIFont systemFontOfSize:fontSizeDispatch + FOUR_FONT]];
    //縁取り
    [[btn layer] setBorderColor:[[UIColor blackColor] CGColor]]; // 枠線の色
    [[btn layer] setBorderWidth:BOARD_WITH]; // 枠線の太さ

    return btn;
}

//viewdidloadで画面描画終了後にこっちで色々ロジック回す
- (void)viewDidAppear:(BOOL)animated
{
    //appDelegateをインスタンス化
    appDelegate = [[UIApplication sharedApplication] delegate];
    
    //appDelegateにフォントサイズを格納
    appDelegate.fontSize = fontSizeDispatch;
    
    //初回起動時以外はデータ取得しない
    if (appDelegate.bootFlg) return;
    
    //appdelegateで通信かDB作成が失敗してたら終了
    if (appDelegate.abnomalFlg) {
        [self alertDisplay:appDelegate.abnomalMessage];
        abnomalFlg = YES;
        abnomalMessage = appDelegate.abnomalMessage;
        return;
    }
    
    //データ取得中はインジケータ回し始める
    [SVProgressHUD showWithStatus:@"データ取得中..." maskType:SVProgressHUDMaskTypeBlack];
    
    //別スレッドで各データの取得処理を行う
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, ZERO), ^{
    
        
        //とりあえずデータ取得猶予を持たせるために3秒スリープ
        [NSThread sleepForTimeInterval:SLEEP_TIME];
        
        //DBインスタンス生成
        ControllDataBase *dbControll = [ControllDataBase new];
        
        //kaisuテーブルにデータがないパターン
        //1.テーブルへのinsert処理が始まってない（テーブルデータ０件）
        //2.テーブルへのinsert処理が終わってない（テーブルデータ？件　※該当する開催回データがない）
        //3.totoが開催してない（テーブルデータ？件　※シーズン中にスポット的　レアケース）
        //3.totoが開催してない（テーブルデータ０件　※シーズンオフ）
        
        //シーズンオフ中はデータinsert時に色づけしたデータをテーブルに格納して、それをベースに例外処理する？？？？
        //現時点では楽天サイトがシーズンオフ中にどうなるかわかんないから
        
        
        //現在開催中のtoto有無判断（開催中じゃなかったらメッセージ表示して終了）
        int abnomalCount = ZERO;
        while (YES) {
            if ([dbControll kaisuDataCountCheck] != ZERO) {
                //kaisuテーブルにinsertが始まってから少し猶予を持たせるためにさらに3秒スリープ
                [NSThread sleepForTimeInterval:SLEEP_TIME];
                if (![dbControll returnKaisaiYesNo]) {
                    [self alertDisplay:NOT_KAISAI_MESSAGE];
                    abnomalFlg = YES;
                    abnomalMessage = NOT_KAISAI_MESSAGE;
                    return;
                }
                break;
            }
            abnomalCount++;
            //ループが１万回まわっても終わらなかったら異常として終了させる
            if (abnomalCount > ROOP_AN_COUNT) {
                [self alertDisplay:ROOP_AN_MESSAGE1];
                abnomalFlg = YES;
                abnomalMessage = ROOP_AN_MESSAGE1;
                return;
            }
        }
        
        //DBにデータがある状態で以降の処理が必要なのでチェックロジックでループ
        NSString *kaisu = nil;
        abnomalCount = ZERO;
        while (YES) {
            kaisu = [dbControll returnKaisaiNow];
            if ([kaisu length] > ZERO) {
                int drewCheck = [dbControll drewCheck:kaisu];
                if (drewCheck == WAKU_COUNT) break;
            }
            abnomalCount++;
            //ループが１万回まわっても終わらなかったら異常として終了させる
            if (abnomalCount > ROOP_AN_COUNT) {
                [self alertDisplay:ROOP_AN_MESSAGE2];
                abnomalFlg = YES;
                abnomalMessage = ROOP_AN_MESSAGE2;
                return;
            }
        }
        
        //現在開催回数
        appDelegate.kaisu = kaisu;
        
        //DBから現在開催回を取得してそれをキーに組み合わせデータ取得　※戻り値データはホームチーム１３チームの後にアウェイチーム１３チームの系２６オブジェクト
        appDelegate.teamArray = [dbControll returnKumiawase];
        
        //DBから現在開催回を取得してそれをキーに組み合わせtoto支持率データ取得　枠順(ホーム、ドロー、アウェイ)の３９オブジェクト
        appDelegate.rateArrayToto = [dbControll returnRate:TOTO];
        
        //DBから現在開催回を取得してそれをキーに組み合わせbODDS支持率データ取得　※まだデータがない時は「--.--」が返ってくる
        appDelegate.rateArrayBook = [dbControll returnRate:BODDS];
        
        //起動フラグをYESにする
        appDelegate.bootFlg = YES;
        
        //データ取得が正常終了したらメインスレッドで以下の処理を行う
        dispatch_async(dispatch_get_main_queue(), ^{
            
            //インジケータ終了
            [SVProgressHUD showSuccessWithStatus:@"データ取得完了！"];
            
            //現在の回数と販売終了日を表示してあげようかね？
            UILabel *kaisuLabel = [[UILabel alloc] initWithFrame:ENDDATE_LABEL_RECT];
            kaisuLabel.backgroundColor = [UIColor clearColor];
            kaisuLabel.textColor = [UIColor blackColor];
            kaisuLabel.numberOfLines = TWO;
            kaisuLabel.textAlignment = NSTextAlignmentCenter;
            kaisuLabel.text = [NSString stringWithFormat:@"対象回：第%@回 toto\n%@", [kaisu substringWithRange:NSMakeRange(ONE,THREE)], [dbControll returnSaleEndDate]];
            [kaisuLabel setFont:[UIFont systemFontOfSize:fontSizeDispatch - THREE]];
            
            [self.view addSubview:kaisuLabel];
        });
    });
}

//ボタンプッシュ時の動作
- (void)btnPush:(UIButton *)button
{
    //異常発生時は警告表示で画面遷移しない
    if (abnomalFlg) {
        [self alertDisplay:abnomalMessage];
        return;
    }

    //それぞれの画面に遷移
    UINavigationController *ngcl;
    if (button.tag == SINGLE) {
        SingleChoiceViewController *scvc = [SingleChoiceViewController new];
        ngcl = [[UINavigationController alloc] initWithRootViewController:scvc];
    } else {
        MultiChoiceViewController *mcvd = [MultiChoiceViewController new];
        ngcl = [[UINavigationController alloc] initWithRootViewController:mcvd];
    }
    [self presentViewController:ngcl animated:YES completion:nil];;
}

//異常の場合はアラート表示
- (void)alertDisplay:(NSString *)message
{
    [[[UIAlertView alloc] initWithTitle:@""
                                message:message
                               delegate:self
                      cancelButtonTitle:@"OK"
                      otherButtonTitles:nil] show];
   
    //インジケータは即終了
    [SVProgressHUD dismiss];
}

- (void)didReceiveMemoryWarning {[super didReceiveMemoryWarning];}

@end

File: TotoRan/newRootViewController.swift
//
//  newRootViewController.swift
//  TotoRan
//
//  Created by 大山 孝 on 2015/12/03.
//  Copyright © 2015年 raksam.com. All rights reserved.
//

import UIKit

class newRootViewController: UIViewController
{
    var abnomalFlg = false //何かしらの異常がある場合は次画面への遷移を抑制する
    var abnomalMessage = "" //何かしらの異常がある場合に次画面遷移を試みた場合に表示するメッセージ
    var appDelegate:AppDelegate = (UIApplication.sharedApplication().delegate as! AppDelegate)//通信とDB失敗データの取得と各種データの格納用
    var fontSizeDispatch: Int32 = ManyNum.ZERO //機種別フォントサイズ
    
    override func viewDidLoad() {
        super.viewDidLoad()

        //背景色は黄色
        view.backgroundColor = UIColor.yellowColor()
        
        //機種を画面サイズ幅で制御 4シリーズと5シリーズは320.0、6は375.0、6+は414.0　※6sシリーズを考慮すっぺ
        if RectMaker.width() <= DisplaySize.FOUR_WIDTH {
            if RectMaker.height() <= DisplaySize.FIVE_HEIGHT {
                fontSizeDispatch = FontSize.FOUR_FONT
            } else {
                fontSizeDispatch = FontSize.FIVE_FONT
            }
        } else if RectMaker.width() >= DisplaySize.SIX_WIDTH && DisplaySize.SIXPLUS_WIDTH > RectMaker.width() {
            fontSizeDispatch = FontSize.SIX_FONT
        } else {
            fontSizeDispatch = FontSize.SIXPLUS_FONT
        }
        
        //ロゴ
        let logo = UIImageView.init(frame: RectMaker.logoRect())
        logo.image = UIImage(named: FILE_NAME.LOGO)
        view.addSubview(logo)
        
        //ボタン
        let singleBtn = makeButton("シングル選択", frame: RectMaker.singleBtnRect(), tag: TAG.SINGLE, color: UIColor.blackColor())
        let multiBtn = makeButton("マルチ選択", frame: RectMaker.multiBtnRect(), tag: TAG.MULTI, color: UIColor.blackColor())
        let confirmBtn = makeButton("支持率確認", frame: RectMaker.confirmBtnRect(), tag: TAG.CONFIRM, color: UIColor.blueColor())
        let reloadBtn = makeButton("データ再取得", frame: RectMaker.reloadBtnRect(), tag: TAG.RELOAD, color: UIColor.redColor())
        view.addSubview(singleBtn)
        view.addSubview(multiBtn)
        view.addSubview(confirmBtn)
        view.addSubview(reloadBtn)
        
        //appDelegateにフォントサイズを格納
        appDelegate.fontSize = fontSizeDispatch
    }
    
    
    //画面描画処理が終わったらこっちで色々処理
    override func viewDidAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        //appdelegateで通信かDB作成が失敗してたら終了
        if appDelegate.abnomalFlg {
            alertDisplay(appDelegate.abnomalMessage)
            abnomalFlg = true;
            abnomalMessage = appDelegate.abnomalMessage;
            return
        }
        
        //初回起動時以外はデータ取得しない ※強制再取得は別
        if appDelegate.bootFlg { return }
        
        //データ再取得ボタン設置のため、取得処理を切り出した
        fetchProcess()
    }
    
    //データ再取得ボタン設置のため、取得処理を切り出したメソッド
    func fetchProcess() {
        //データ取得中はインジケータ回し始める
        SVProgressHUD.showWithStatus(MESSAGE_STR.DATA_GETING, maskType: .Black)
        
        //DBインスタンス生成
        let dbControll = ControllDataBase()
        
        //開催データの取得判断
        if !dbControll.kaisaiDataCheck() {
            let kaisaiData = GetKaisu()
            kaisaiData.returnSourceString(URL_STR.KAISAI_URL)
            _ = GetTeamNames()  //結果が不要な場合は「_ = 」をつける　気持ち悪い
        }
        
        //支持率データの取得判断
        if let judgeData = dbControll.rateUpdateJudge() {
            let toto = GetRateToto()
            toto.parseRate(judgeData)
        }
        
        //別スレッドで各データの取得処理を行う
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), {
            
            //3秒スリープ
            NSThread.sleepForTimeInterval(ManyNum.SLEEP)
            
            //現在開催中のtoto有無判断（開催中じゃなかったらメッセージ表示して終了）
            var abnomalCount = ManyNum.ZERO
            
            while true {
                if dbControll.kaisuDataCountCheck() != ManyNum.ZERO && dbControll.abbNameDataCountCheck() >= 50 { //少なくともabbnameテーブルの件数が50件以上あるかどうか？（J1+J2+J3だと50は上回る）
                    NSThread.sleepForTimeInterval(ManyNum.SLEEP)
                    if dbControll.returnKaisaiYesNo() == ManyNum.ZERO {
                        //クロージャの外側メソッドを呼ぶのでselfを付与
                        self.alertDisplay(MESSAGE_STR.NOT_KAISAI_MESSAGE)
                        self.abnomalFlg = true
                        self.abnomalMessage = MESSAGE_STR.NOT_KAISAI_MESSAGE
                        //インジケータ終了
                        SVProgressHUD.showSuccessWithStatus("")
                        return
                    }
                    break
                }
                abnomalCount++
                //ループが１万回まわっても終わらなかったら異常として終了させる
                if abnomalCount > ManyNum.ROOP_AN_COUNT {
                    //クロージャの外側メソッドを呼ぶのでselfを付与
                    self.alertDisplay(MESSAGE_STR.ROOP_AN_MESSAGE1)
                    self.abnomalFlg = true
                    self.abnomalMessage = MESSAGE_STR.ROOP_AN_MESSAGE1
                    //インジケータ終了
                    SVProgressHUD.showSuccessWithStatus("")
                    return
                }
            }
            
            //DBにデータがある状態(13枠分のデータ)で以降の処理が必要なのでチェックロジックでループ
            abnomalCount = ManyNum.ZERO
            while true {
                if let kaisu = dbControll.returnKaisaiNow() {
                    if dbControll.drewCheck(kaisu) == ManyNum.WAKU_COUNT {
                        //現在開催回数だけはわざわざ変数宣言して格納する必要がないようにここで格納しとく
                        self.appDelegate.kaisu = kaisu
                        break
                    }
                }
                abnomalCount++
                //ループが１万回まわっても終わらなかったら異常として終了させる
                if abnomalCount > ManyNum.ROOP_AN_COUNT {
                    //クロージャの外側メソッドを呼ぶのでselfを付与
                    self.alertDisplay(MESSAGE_STR.ROOP_AN_MESSAGE2)
                    self.abnomalFlg = true
                    self.abnomalMessage = MESSAGE_STR.ROOP_AN_MESSAGE2
                    //インジケータ終了
                    SVProgressHUD.showSuccessWithStatus("")
                    return
                }
                
            }
            
            //DBから現在開催回を取得してそれをキーに組み合わせデータ取得　※戻り値データはホームチーム１３チームの後にアウェイチーム１３チームの系２６オブジェクト
            self.appDelegate.teamArray = dbControll.returnKumiawase()
            
            //DBから現在開催回を取得してそれをキーに組み合わせtoto支持率データ取得　枠順(ホーム、ドロー、アウェイ)の３９オブジェクト
            self.appDelegate.rateArrayToto = dbControll.returnRate(ODDS.TOTO)
            
            //DBから現在開催回を取得してそれをキーに組み合わせbODDS支持率データ取得　※まだデータがない時は「--.--」が返ってくる
            self.appDelegate.rateArrayBook = dbControll.returnRate(ODDS.BOOK)
            
            //起動フラグをYESにする
            self.appDelegate.bootFlg = true
            
            //メインスレッドに処理を戻す
            dispatch_async(dispatch_get_main_queue(), {
                
                //インジケータ終了
                SVProgressHUD.showSuccessWithStatus(MESSAGE_STR.DATA_GETED)
                
                //現在の回数と販売終了日を表示してあげようかね？
                let kaisuLabel = UILabel(frame: RectMaker.kaisuLabelRect())
                kaisuLabel.backgroundColor = UIColor.clearColor()
                kaisuLabel.textColor = UIColor.blackColor()
                kaisuLabel.numberOfLines = 2
                kaisuLabel.textAlignment = .Center
                kaisuLabel.text = "対象回：第\((self.appDelegate.kaisu as NSString).substringFromIndex(1))回 toto\n\(dbControll.returnSaleEndDate())"
                kaisuLabel.font = UIFont.systemFontOfSize(CGFloat(self.fontSizeDispatch - FontSize.SYSTEMSIZE))
                self.view.addSubview(kaisuLabel)
            })
        })
    }

    //ボタン生成メソッド
    func makeButton(title: String, frame: CGRect, tag: Int, color: UIColor) -> UIButton
    {
        let btn = UIButton(frame: frame)
        btn.tag = tag
        btn.setTitleColor(color, forState: .Normal)
        btn.setTitleColor(UIColor.lightGrayColor(), forState: .Highlighted)
        btn.setTitle(title, forState: .Normal)
        btn.addTarget(self, action: "btnPush:", forControlEvents: .TouchUpInside)
        btn.titleLabel?.font = UIFont.systemFontOfSize(CGFloat(fontSizeDispatch + FontSize.FOUR_FONT))
        
        //縁取り
        btn.layer.borderColor = color.CGColor
        btn.layer.borderWidth = DisplaySize.BOARD_WITH
        
        return btn
    }
    
    //ボタンプッシュアクションメソッド
    func btnPush(button: UIButton) {
        //異常発生時は警告表示で画面遷移しない
        if abnomalFlg {
            alertDisplay(abnomalMessage)
            return
        }
        
        //それぞれの画面に遷移かデータ再取得
        var ngcl: UINavigationController!
        if button.tag == TAG.SINGLE {
            let scvc = SingleChoiceViewController()
            ngcl = UINavigationController(rootViewController: scvc)
        } else if button.tag == TAG.MULTI {
            let mcvc = MultiChoiceViewController()
            ngcl = UINavigationController(rootViewController: mcvc)
        } else if button.tag == TAG.CONFIRM {
            let rcvc = RateConfirmViewController()
            rcvc.modalTransitionStyle = .FlipHorizontal
            ngcl = UINavigationController(rootViewController: rcvc)
        } else if button.tag == TAG.RELOAD {
            //DBインスタンス生成してkaisuテーブル、kumiawaseテーブルのデータ削除
            let dbControll = ControllDataBase()
            dbControll.deleteTables()
            
            //データを再取得
            fetchProcess()
            
            //画面遷移はなし
            return
        }
        
        self.presentViewController(ngcl, animated: true, completion: nil)
    }
    
    //アラートメッセージ表示メソッド
    func alertDisplay(message: String) {
        let alertController = UIAlertController(title: "", message: message, preferredStyle: .Alert)
        
        let defaultAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
        alertController.addAction(defaultAction)
        
        presentViewController(alertController, animated: true, completion: nil)
        
    }
}
